sqlglot/parser.py

```diff
@@ -5475,6 +5476,37 @@ def _parse_comprehension(
             condition=condition,
         )
 
+    def _parse_heredoc(self) -> exp.RawString:
+        # Note: we've already matched a $ token before reaching here
+        self._match(TokenType.PARAMETER)
+        tag_start = self._prev
+
+        while not self._match(TokenType.PARAMETER):
+            self._advance()
+            if not self._curr:
+                self.raise_error("No closing $ found.")
+
+        tag_text = self._find_sql(tag_start, self._prev)
+        heredoc_start = self._curr
+        heredoc_end = self._prev
+
+        tag_start = None
+        while True:
+            if self._match(TokenType.PARAMETER, advance=False):
+                if tag_start and self._find_sql(tag_start, self._curr) == tag_text:
+                    this = self._find_sql(heredoc_start, heredoc_end)
+                    self._advance()
```

Comment:
Can this be simplified? I'd expect this section to just contain some lookahead checks. For example: "if there's a $ followed by a token that matches the tag text (if any), _and_ there's another dollar, we're at the heredoc's end and only then we should capture the raw text with `_find_sql`".